client/ayon_core/host/interfaces/workfiles.py

```diff
@@ -0,0 +1,1214 @@
+from __future__ import annotations
+
+import os
+import platform
+import shutil
+import typing
+from abc import abstractmethod
+from dataclasses import dataclass, asdict
+from typing import Optional, Any
+
+import ayon_api
+import arrow
+
+from ayon_core.lib import emit_event
+
+if typing.TYPE_CHECKING:
+    from ayon_core.pipeline import Anatomy
+
+
+WORKFILE_OPEN_REASON = "workfile.opened"
+WORKFILE_SAVE_REASON = "workfile.saved"
+
+
+@dataclass
+class WorkfileInfo:
+    """Information about workfile.
+
+    Host can open, copy and use the workfile using this information object.
+
+    Attributes:
+        filepath (str): Path to the workfile.
+        rootless_path (str): Path to the workfile without root. And without
+            backslashes on Windows.
+        file_size (Optional[float]): Size of the workfile in bytes.
+        file_created (Optional[float]): Timestamp when the workfile was
+            created on the filesystem.
+        file_modified (Optional[float]): Timestamp when the workfile was
+            modified on the filesystem.
+        workfile_entity_id (Optional[str]): Workfile entity id. If None then
+            the workfile is not in the database.
+        description (str): Description of the workfile.
+        created_by (Optional[str]): User id of the user who created the
+            workfile entity.
+        updated_by (Optional[str]): User id of the user who updated the
+            workfile entity.
+        available (bool): True if workfile is available on the machine.
+
+    """
+    filepath: str
+    rootless_path: str
+    file_size: Optional[float]
+    file_created: Optional[float]
+    file_modified: Optional[float]
+    workfile_entity_id: Optional[str]
+    description: str
+    created_by: Optional[str]
+    updated_by: Optional[str]
+    available: bool
+
+    @classmethod
+    def new(
+        cls,
+        filepath: str,
+        rootless_path: str,
+        available: bool,
+        workfile_entity: dict[str, Any],
+    ):
+        file_size = file_modified = file_created = None
+        if filepath and os.path.exists(filepath):
+            filestat = os.stat(filepath)
+            file_size = filestat.st_size
+            file_created = filestat.st_ctime
+            file_modified = filestat.st_mtime
+
+        if workfile_entity is None:
+            workfile_entity = {}
+
+        attrib = {}
+        if workfile_entity:
+            attrib = workfile_entity["attrib"]
+
+        return cls(
+            filepath=filepath,
+            rootless_path=rootless_path,
+            file_size=file_size,
+            file_created=file_created,
+            file_modified=file_modified,
+            workfile_entity_id=workfile_entity.get("id"),
+            description=attrib.get("description") or "",
+            created_by=workfile_entity.get("createdBy"),
+            updated_by=workfile_entity.get("updatedBy"),
+            available=available,
+        )
+
+    def to_data(self) -> dict[str, Any]:
```

Comment:
This can make use of *TypedDict* because you know exactly what data you need to build `WorkfileInfo` obejct. Basically `TypedDict` with signature from the constructor. The same applies to `from_data()` below.